actors/ppo_atari_pz.py

```diff
@@ -0,0 +1,1034 @@
+        observation_space = environment_specs.get_observation_space()
+        action_space = environment_specs.get_action_space(seed=config.seed)
+
+        # Get model
+        model, _, version_info = await actor_session.model_registry.retrieve_version(
+            PPOModel, config.model_id, config.model_version
+        )
+        log.info(f"Actor - retreved model number: {version_info['version_number']}")
+        obs_shape = model.input_shape[::-1]
+
+        async for event in actor_session.all_events():
+            if event.observation and event.type == cogment.EventType.ACTIVE:
+                if (
+                    event.observation.observation.HasField("current_player")
+                    and event.observation.observation.current_player != actor_session.name
+                ):
+                    # Not the turn of the agent
+                    actor_session.do_action(action_space.serialize(action_space.create()))
+                    continue
+                obs = observation_space.deserialize(event.observation.observation)
+                obs_tensor = torch.tensor(obs.flat_value, dtype=self._dtype).reshape(obs_shape)
+                obs_tensor = torch.unsqueeze(obs_tensor.permute((2, 0, 1)), dim=0).to(torch.device(model.device))
+
+                # Get action from policy network
+                with torch.no_grad():
+                    dist = model.network.get_action(obs_tensor)
+                    action = dist.sample().cpu().numpy()[0]
+
+                # Send action to environment
+                assert isinstance(action_space.gym_space, Discrete)  # TODO: test with other action space types
+                action_value = action_space.create(value=action)
+                actor_session.do_action(action_space.serialize(action_value))
+
+
+class BasePPOTraining(ABC):
+    """Abstract class for training PPO agent"""
+
+    default_cfg = {
+        "seed": 0,
+        "num_epochs": 10,
+        "num_iter": 500,
+        "epoch_num_trials": 1,
+        "num_parallel_trials": 1,
+        "discount_factor": 0.99,
+        "entropy_loss_coef": 0.05,
+        "value_loss_coef": 0.5,
+        "action_loss_coef": 1.0,
+        "clipping_coef": 0.1,
+        "learning_rate": 3e-4,
+        "batch_size": 64,
+        "num_steps": -1,  # End of trial
+        "lambda_gae": 0.95,
+        "device": "cpu",
+        "grad_norm": 0.5,
+        "lr_decay_factor": 0.999,
+        "image_size": [6, 84, 84],
+        "buffer_capacity": 100_000,
+    }
+
+    def __init__(self, environment_specs: EnvironmentSpecs, cfg: EnvironmentConfig) -> None:
+        super().__init__()
+        self._dtype = torch.float32
+        self._environment_specs = environment_specs
+        self._cfg = cfg
+        self._device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        self.returns = 0
+
+        # Set random seed
+        torch.manual_seed(self._cfg.seed)
```

Comment:
Given the async nature of what we do this should be called before each operation that uses randomness (seeding it with a random number from the main rng)

Reply:
Done